File: bots/node/lib/bot.js
var express = require( 'express' );
var http = require( 'http' );
var assert = require( 'assert' );
var bodyParser = require( 'body-parser' );
var multer = require( 'multer' );

var app = express();

app.use( bodyParser.urlencoded( {
    extended: true
} ) );

app.use( multer( {
    dest: './uploads/'
} ) );

app.listen( 8000 );

app.get( '/bot', function( req, res ) {
    res.json( {
        botname: 'sample_botname',
        version: '0.1.0',
        username: 'sample_username'
    } );
} );

app.post( '/game', function( req, res ) {
    res.json( {} );
} );

app.post( '/round', function( req, res ) {
    function randomDirection( array ) {
        var directions = [ 'NORTH', 'EAST', 'SOUTH', 'WEST' ];
        return directions[ array[ Math.floor( Math.random() * array.length ) ] ];
    }

    var round = req.body.round;
    var map = JSON.parse( req.body.map );
    var gameid = req.body.gameid;
    var myid = req.body.myid;
    var W = req.body.W;
    var H = req.body.H;
    var intent = [];

    for ( var i = 0, len = map.length; i < len; ++i ) {
        var creature = map[ i ];
        if ( creature.userid == myid && creature.hp > 0 ) {
            var x = creature.x, y = creature.y, hp = creature.hp;
            var attacks = [], moves = [];
            var offsets = [
                [ 0, 1 ],
                [ 1, 0 ],
                [ 0, -1 ],
                [ -1, 0 ]
            ];

            for ( var j = 0; j < offsets.length; ++j ) {
                var newX = parseInt( x, 10 ) + offsets[ j ][ 0 ], newY = parseInt( y, 10 ) + offsets[ j ][ 1 ];

                if ( newX >= 0 && newX < W && newY >= 0 && newY < H ) {
                    var creatureFound = false;

                    for ( var z = 0; z < len; ++z ) {
                        if ( map[ z ].x == newX && map[ z ].y == newY ) {
                            creatureFound = true;

                            if ( map[ z ].userid != myid && map[ z ].hp > 0 ) {
                                attacks.push( j );
                            }
                        }
                    }

                    if ( !creatureFound ) {
                        moves.push( j );
                    }
                }
            }

            if ( attacks.length ) {
                intent.push( {
                    'creatureid': creature.creatureid,
                    'action': 'ATTACK',
                    'direction': randomDirection( attacks )
                } );
            }
            else {
                if ( !moves.length || Math.random() < 0.5 ) {
                    intent.push( {
                        'creatureid': creature.creatureid,
                        'action': 'NONE',
                        'direction': 'NONE'
                    } );
                }
                else {
                    intent.push( {
                        'creatureid': creature.creatureid,
                        'action': 'MOVE',
                        'direction': randomDirection( moves )
                    } );
                }
            }
        }
    }

    res.json( {
        'intent': intent
    } );
} );

module.exports = app;

File: bots/python/bot.py
import json, random, sys
from urlparse import parse_qs
from bottle import route, run, template, request, response, hook

@hook('after_request')
def contentType():
    response.headers['Content-type'] = 'application/json'
    
@route('/bot', method='GET')
def bot():
    return json.dumps({
        'botname': 'pythonbot',
        'version': '0.1.0',
        'username': sys.argv[1]
    })

@route('/game', method='POST')
def game():
    return json.dumps([])

@route('/round', method='POST')
def round():
    vars = {
        'map': request.forms.get('map'),
        'gameid': request.forms.get('gameid'),
        'W': request.forms.get('W'),
        'H': request.forms.get('H'),
        'round': request.forms.get('round'),
        'myid': request.forms.get('myid')
    }

    map = json.loads(vars['map'])
    game_id, my_id = vars['gameid'], vars['myid']
    round, W, H = vars['round'], vars['W'], vars['H']

    my_creatures = [creature for creature in map if creature['userid'] == int(my_id)]
    enemy_creatures = [creature for creature in map if creature['userid'] != int(my_id)]

    occupied = set([creature['x'] + creature['y'] * 1j for creature in map])
    enemy_occupied = set([creature['x'] + creature['y'] * 1j for creature in enemy_creatures])

    intents = []
    for my_creature in my_creatures:
        possible_moves = set()
        possible_attacks = set()
        intent = {'creatureid': my_creature['creatureid']}

        location = my_creature['x'] + my_creature['y'] * 1j
        directions = {'north': -1j, 'east': 1, 'south': 1j, 'west': -1}

        for direction, offset in directions.iteritems():
            moved = location + offset
            if 0<= moved.real < W and 0 <= moved.imag < H:
                if moved not in occupied:
                    possible_moves.add(direction)
                if moved in enemy_occupied:
                    possible_attacks.add(direction)
        if possible_attacks:
            intent['action'] = 'attack'
            intent['direction'] = random.choice(list(possible_attacks))
        elif possible_moves and random.randint(0, 1) == 0:
            intent['action'] = 'move'
            intent['direction'] = random.choice(list(possible_moves))

        if 'action' in intent:
            intents.append(intent)

    return json.dumps({'intent': intents})

if len(sys.argv) != 3:
    print('Syntax: python bot.py <username> <port>')
    sys.exit(0)

run(host='localhost', port=sys.argv[2], reloader=True)
